### ISSUE TITLE
Unit Testing Components

### ISSUE THREAD

**Issue by BE-Webdesign** [WordPress/gutenberg#641]
Would unit testing components be of value to the project?  I feel somewhat like I am blocking up the flow of progress. It doesn't appear that anyone is pursuing testing too much, maybe that would be a good place for me to fit in and be useful?  Negatives is it will slow down our CI build and as things change it will slow down PRs, but we could do the component testing separately and eventually streamline it into the CI so we don't have those kind of project flow problems until things are more stable.

**Comment by nylen** [comment]
> It doesn't appear that anyone is pursuing testing too much

I'd still like to see a rule that we don't merge new code until it has tests.  Generally I'm of the opinion that the more automated testing, the better.  In fact I've considered changing my name to "James `needs-unit-tests` Nylen".

Unit-testing components (and refactoring them so that they can be more easily tested) would be extremely valuable and we would love your help with it.  Build time should not be a consideration in making this decision.

Integration testing of actual flows would also be very valuable; however, these tests do slow things down quite a bit.  Prior art from Calypso:  https://github.com/Automattic/wp-e2e-tests

**Comment by aduth** [comment]
What did you have in mind for component testing? What sorts of behaviors should we be testing?

I'm very open toward something like [Enzyme](https://github.com/airbnb/enzyme), with an emphasis on shallow render testing. I don't think there's a performance worry here (per your note on slower CI build).

We already create a fake DOM context in our test bootstrap:

https://github.com/WordPress/gutenberg/blob/master/bootstrap-test.js

**Comment by BE-Webdesign** [comment]
Yeah, I was thinking of using Enzyme.  https://github.com/Automattic/wp-e2e-tests is a lot to take in 😅.

>What sorts of behaviors should we be testing?

I think testing rendering with default state, and basic state changes is a good place to start. Testing the editor components as well first would be good as they are not changing as frequently as the block components.  That way as the block API changes we will know whether or not our components are breaking at least, (hopefully).  I don't know if we will want to use the Full DOM as well to keep track of focusing.

**Comment by BE-Webdesign** [comment]
Going to breathe some life back into this endeavor, after I touch up meta boxes.

**Comment by gziolo** [comment]
I updated the list of all components with the actual status. I omitted `edit-post` which is going to be moved into its own folder soon. This list is huge. @mtias do we want to keep this issue open?
